Sprint _15

MiraiChan opened the pull request:
Sprint _15 done.

MiraiChan committed:
ScrollView, UICollectionViews for Color and Emoji sets were added.
Constraints were activated.
Method createButtonTapped supplemented.

MiraiChan committed:
UICollectionViews for color and emoji were configured for NewHabitViewController:
DataSourse and DelegateFlowLayout were added and set.
According Cells and Headers were added and set up.

MiraiChan committed:
IrregularEventViewController was fully supplemented with UICollectionViews for color and emoji.
DataSourse and DelegateFlowLayout were added and set.
According Cells and Headers were added and set up.

MiraiChan committed:
*.xcdatamodeld models were created:
TrackerCoreData, TrackerCategoryCoreData, TrackerRecordCoreData.
Relationship b/n entities was set.

MiraiChan committed:
Type of relationship was set to "to many".
Optional properties were removed.

MiraiChan committed:
NSPersistentContainer was initialized.

MiraiChan committed:
New classes as data layers were created:
TrackerStore, TrackerCategoryStore, TrackerRecordStore.

MiraiChan committed:
Protocols, according constants (to get data changes from Core Data) and class init-s were added to TrackerStore and TrackerRecordStore.

MiraiChan committed:
NSFetchedResultsController is used to get updates from Core Data.
Store files were configured.

MiraiChan committed:
fatalErrors were replaced with assertionFailures.

MiraiChan committed:
Delegates were set for the trackerStore and trackerRecordStore.
Trackers and completed trackers were retrieved from their respective stores.
Stubs were added to conform protocol TrackerRecordStoreDelegate.
Method incompleteTracker was renewed.

MiraiChan committed:
Context issue fixed.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -6,18 +6,12 @@
 //
 
 import UIKit
+import CoreData
 
```
Часто в разработке под iOS(и не только) используют [Realm](https://realm.io/realm-swift/). Сам я уже много лет не используют `CoreData`.

MiraiChan replied:
Проверьте мне работу, пожалуйстааааа...26 часов назад сдала -_-''

Helvetii951 commented on Tracker/CoreData/CoreDataHelper/UIColorMarshalling.swift:
```diff
@@ -0,0 +1,35 @@
+//Алгоритмы сериализации и десериализации цвета
+import UIKit
+
+final class UIColorMarshalling {
+    func hexString(from color: UIColor) -> String {
+        let components = color.cgColor.components
+        let r: CGFloat = components?[0] ?? 0.0
+        let g: CGFloat = components?[1] ?? 0.0
+        let b: CGFloat = components?[2] ?? 0.0
+        return String.init(
+            format: "%02lX%02lX%02lX",
+            lroundf(Float(r * 255)),
+            lroundf(Float(g * 255)),
+            lroundf(Float(b * 255))
+        )
+    }
+    
+    func color(from hex: String) -> UIColor? {
+        var rgbValue:UInt64 = 0
+        Scanner(string: hex).scanHexInt64(&rgbValue)
+        return UIColor(
+            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
+            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
+            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
+            alpha: CGFloat(1.0)
+        )
+    }
+}
```
Функции можно сделать статическими Так на не нужно будет создавать экземпляр `UIColorMarshalling`. Создание экземпляров классов - не самая дешевая операция.

MiraiChan replied:
Ok

Helvetii951 commented on Tracker/DataMock/TrackersStorage.swift:
```diff
@@ -0,0 +1,53 @@
+
+import UIKit
+
+final class TrackersStorage {
+    static let shared = TrackersStorage()
+    var categories: [TrackerCategory] = [
```
Часто при использовании `static let shared` используют приватный инициализатор. Тем самым реализуюя паттерн [singleton](https://theswiftdev.com/swift-singleton-design-pattern/).

MiraiChan replied:
Угу

Helvetii951 commented on Tracker/DataMock/TrackersStorage.swift:
```diff
@@ -0,0 +1,53 @@
+//            ]
+//        ),
+//        
+//        TrackerCategory(
+//            title: "Радостные мелочи",
+//            trackers: [
+//                Tracker (
+//                    id: UUID, name: "Кошка заслонила камеру на созвоне",
+//                    color: .ypColorSelection12,
+//                    emoji:"😻",
+//                    schedule: [TrackerSchedule.DaysOfTheWeek.saturday]
+//                ),
```
Комменты с тестовыми данными лучше сразу убирать.

MiraiChan replied:
Хорошо

Helvetii951 commented on Tracker/TrackersViewController/AddingTracker/NewTrackerViewController/IrregularEvent/EventColorCell.swift:
```diff
@@ -0,0 +1,40 @@
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        
+        colorView.frame = CGRect(
+            x: (contentView.bounds.width - 40) / 2,
+            y: (contentView.bounds.height - 40) / 2,
+            width: 40,
+            height: 40
+        )
```
`40`, `2`. Это - [магические числа.](https://bureau.ru/soviet/20230406/). Желательно выносить эти значения в свойства или структуры с понятным названием.

MiraiChan replied:
Done.

Helvetii951 commented on Tracker/TrackersViewController/AddingTracker/NewTrackerViewController/IrregularEvent/EventEmojiCell.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    static var reuseId = "Event emoji cell"
+    
+    let emojiLabel: UILabel = {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.font = UIFont.systemFont(ofSize: 32)
+        label.textAlignment = .center
+        return label
+    }()
```
Рекомендую всегда использовать [ленивые свойства ](https://abhimuralidharan.medium.com/lazy-var-in-ios-swift-96c75cb8a13a)для объявления подобных свойств.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.

MiraiChan replied:
Done.

Helvetii951 commented on Tracker/CoreData/DataStore/TrackerStore.swift:
```diff
@@ -0,0 +1,121 @@
+        return trackers
+    }
+    
+    convenience override init() {
+        guard let context = (UIApplication.shared.delegate as? AppDelegate)?.context else {
+            fatalError("Failed to obtain the Core Data context.")
+        }
+        try! self.init(context: context)
+    }
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfailure(_:file:line:))

MiraiChan replied:
Done.

MiraiChan committed:
Mock file was deleted due to further uselessness.
Method reloadData was changed because of reference to the deleted file.
"Static" and "lazy" were added, magic numbers were replaced with constants.
fatalError replaced with assertionFailure.

MiraiChan committed:
Displaying a row of selected days in the "Schedule" cell was added.

MiraiChan committed:
The creation of UIColorMarshalling class instances has been eliminated.

MiraiChan committed:
Set the createButton logic.
Added new method to check the state of a button.
